# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieCommons.cpp
# Commit: d10ae1bb47a8
# Full Hash: d10ae1bb47a8b1cc81718ace05688e97d583e27f
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2025-01-10 21:44:02
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r 336214ad7891 -r d10ae1bb47a8 netwerk/cookie/CookieCommons.cpp
--- a/netwerk/cookie/CookieCommons.cpp	Fri Jan 10 08:24:29 2025 +0000
+++ b/netwerk/cookie/CookieCommons.cpp	Fri Jan 10 08:24:30 2025 +0000
@@ -424,9 +424,14 @@
 
   nsCString cookieString(aCookieString);
 
+  nsCOMPtr<nsILoadInfo> loadInfo =
+      aDocument->GetChannel() ? aDocument->GetChannel()->LoadInfo() : nullptr;
+  const bool on3pcbException = loadInfo && loadInfo->GetIsOn3PCBExceptionList();
+
   aCookieParser.Parse(baseDomain, requireHostMatch, cookieStatus, cookieString,
                       EmptyCString(), false, isForeignAndNotAddon,
-                      mustBePartitioned, aDocument->IsInPrivateBrowsing());
+                      mustBePartitioned, aDocument->IsInPrivateBrowsing(),
+                      on3pcbException);
 
   if (!aCookieParser.ContainsCookie()) {
     return nullptr;
@@ -511,7 +516,8 @@
 bool CookieCommons::ShouldIncludeCrossSiteCookie(Cookie* aCookie,
                                                  bool aPartitionForeign,
                                                  bool aInPrivateBrowsing,
-                                                 bool aUsingStorageAccess) {
+                                                 bool aUsingStorageAccess,
+                                                 bool aOn3pcbException) {
   MOZ_ASSERT(aCookie);
 
   int32_t sameSiteAttr = 0;
@@ -519,15 +525,14 @@
 
   return ShouldIncludeCrossSiteCookie(
       sameSiteAttr, aCookie->IsPartitioned() && aCookie->RawIsPartitioned(),
-      aPartitionForeign, aInPrivateBrowsing, aUsingStorageAccess);
+      aPartitionForeign, aInPrivateBrowsing, aUsingStorageAccess,
+      aOn3pcbException);
 }
 
 // static
-bool CookieCommons::ShouldIncludeCrossSiteCookie(int32_t aSameSiteAttr,
-                                                 bool aCookiePartitioned,
-                                                 bool aPartitionForeign,
-                                                 bool aInPrivateBrowsing,
-                                                 bool aUsingStorageAccess) {
+bool CookieCommons::ShouldIncludeCrossSiteCookie(
+    int32_t aSameSiteAttr, bool aCookiePartitioned, bool aPartitionForeign,
+    bool aInPrivateBrowsing, bool aUsingStorageAccess, bool aOn3pcbException) {
   // CHIPS - If a third-party has storage access it can access both it's
   // partitioned and unpartitioned cookie jars, else its cookies are blocked.
   //
@@ -538,7 +543,7 @@
        (aInPrivateBrowsing &&
         StaticPrefs::
             network_cookie_cookieBehavior_optInPartitioning_pbmode())) &&
-      !aCookiePartitioned && !aUsingStorageAccess) {
+      !aCookiePartitioned && !aUsingStorageAccess && !aOn3pcbException) {
     return false;
   }
 